distributed/client.py

```diff
@@ -3390,6 +3390,34 @@ def get_task_stream(self, start=None, stop=None, count=None):
         return self.sync(self.scheduler.get_task_stream, start=start,
                          stop=stop, count=count)
 
+    @gen.coroutine
+    def _register_worker_setup(self, function, *args, **kwargs):
+        responses = yield self.scheduler.register_worker_setup(function=dumps(function),
+                                                           args=dumps(args),
+                                                           kwargs=dumps(kwargs))
+        results = {}
+        for key, resp in responses.items():
+            if resp['status'] == 'OK':
+                results[key] = resp['result']
+            elif resp['status'] == 'error':
+                six.reraise(*clean_exception(**resp))
+        raise gen.Return(results)
+
+    def register_worker_setup(self, function, *args, **kwargs):
```

Comment:
Actually, I had to add this in order to be able to make things work with `dask_worker` optional keyword, which is filled on worker side. I don't really know why this is needed, as you say, the initial intent on client side is to only pass the function, and we don't register argument. So if someone can explain me what to do here, please do!

Reply:
I can't think of a use case for the function taking arguments beyond the worker, and that complicates things (perhaps unnecessarily). If it's truly meant to be a setup/teardown function, it should have state fully enclosed. I'd prefer the setup functions be forced to have the signature `setup()` or `setup(dask_worker)`, rather than arbitrary signatures

I read

> Actually, I had to add this in order to be able to make things work with dask_worker optional keyword, which is filled on worker side.

as meaning the args/kwargs were needed somehow to play well with distributed's internals. I'm not sure why/how that would be, @mrocklin would know more.

Reply:
OK, I'm fine with keeping things constrained to just the single possible argument.  I'm also ok with forcing it as the input (all setup functions must accept a single argument).  

It's easy to expand these things later if we find that they're necessary.

Reply:
Not sure what the conclusion is here. Force dask_worker arg in fonction signature and no other args? Currently no args are stored in the scheduler context.

How should the register_worker_setup function signature look like on client side?

Reply:
I think that the decisions is to not expose `*args` or `**kwargs` in the client or scheduler code yet.  

We should check the function for the `dask_worker` parameter in the worker code and apply it there if it exists.